### PR TITLE
Transcripts: Display speaker info

### PR DESCRIPTION
**Pull request by ashiagr**
## Description
This displays speaker info for transcripts.

## Testing Instructions
1. Load an SRT format transcript with speaker info (e.g. from `Cautionary Tales`)
2. ✅ Notice that speaker info is displayed properly
3. Prioritize order for VTT format in `TranscriptsManagerImpl.supportedFormats` at line 22
4. Install the app and load a VTT format transcript with speaker info (e.g. from `Build your Saas`)
5. ✅ Notice that speaker info is displayed properly
6. Search for speaker name
7. ✅ Notice that speaker names are included in search results and highlighted correctly

## Screenshots or Screencast 

SRT | VTT 
---|---
 | 

## Checklist
- [x] ~~If this is a user-facing change, I have added an entry in CHANGELOG.md~~
- [x] Ensure the linter passes (`./gradlew spotlessApply` to automatically apply formatting/linting)
- [x] I have considered whether it makes sense to add tests for my changes
- [x] ~~All strings that need to be localized are in `modules/services/localization/src/main/res/values/strings.xml`~~
- [x] Any jetpack compose components I added or changed are covered by compose previews
- [x] ~~I have updated (or requested that someone edit) [the spreadsheet](https://docs.google.com/spreadsheets/d/107jqrutZhU0fVZJ19SBqxxVKbV2NWSdQC9MFYdLiAxc/edit?usp=sharing) to reflect any new or changed analytics.~~
 
#### I have tested any UI changes...

- [x] ~~with different themes~~
- [x] ~~with a landscape orientation~~
- [x] ~~with the device set to have a large display and font size~~
- [x] ~~for accessibility with TalkBack~~

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          7 Errors
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please add tests for class <code>Cue</code> (or add <code>unit-tests-exemption</code> label to ignore this).</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please add tests for class <code>Builder</code> (or add <code>unit-tests-exemption</code> label to ignore this).</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please add tests for class <code>WebvttCueParser</code> (or add <code>unit-tests-exemption</code> label to ignore this).</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please add tests for class <code>WebvttCueInfoBuilder</code> (or add <code>unit-tests-exemption</code> label to ignore this).</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please add tests for class <code>StyleMatch</code> (or add <code>unit-tests-exemption</code> label to ignore this).</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please add tests for class <code>StartTag</code> (or add <code>unit-tests-exemption</code> label to ignore this).</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please add tests for class <code>Element</code> (or add <code>unit-tests-exemption</code> label to ignore this).</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: modules/features/player/src/main/java/androidx/media3/extractor/text/webvtt/WebvttCueParser.java [hunk 1]
```diff
@@ -0,0 +1,998 @@
...
+         line = webvttData.readLine()) {
+      if (textBuilder.length() > 0) {
+        textBuilder.append("\n");
+      }
+      textBuilder.append(line.trim());
+    }
+    String markup = textBuilder.toString();
+    builder.text = parseCueText(id, markup, styles);
+    // parseCueText strips tags like TAG_VOICE "v" and set it on the cue text. This sets original markup wih tags in cue to allow extracting it in the application.
+    builder.markup = markup;
```
Comment (ashiagr) [inline modules/features/player/src/main/java/androidx/media3/extractor/text/webvtt/WebvttCueParser.java:369]: `Cue.text` for vtt format parser strips out tags like TAG_VOICE "v" without setting it on the cue. This sets original markup with tags on the cue to allow extracting any tag in the application in 7548a3f.

I copied original source code for `WebvttCueParser` and `Cue` to make the change as the classes were final and could not be subclassed. Curious to know if there's a better way to handle this.

Comment (MiSikora) [reply modules/features/player/src/main/java/androidx/media3/extractor/text/webvtt/WebvttCueParser.java:369]: Could you report the issue in the `media` repository? Hopefully they can add needed APIs and we can remove copied classes in the future. If they agree that this is something they could add we could make a PR to their repo.

Comment (ashiagr) [reply modules/features/player/src/main/java/androidx/media3/extractor/text/webvtt/WebvttCueParser.java:369]: Asked here: https://github.com/androidx/media/issues/1632

File: modules/features/player/src/main/java/au/com/shiftyjelly/pocketcasts/player/view/transcripts/TranscriptViewModel.kt [hunk 2]
```diff
@@ -152,15 +147,42 @@ class TranscriptViewModel @Inject constructor(
...
             cuesWithTimingSubtitle
                 .flatMap { it.cues }
-                .forEach { cue -> append(cue.text) }
+                .forEach { cue ->
+                    // Extract speaker
+                    val markupText = if (transcriptFormat == TranscriptFormat.VTT) cue.markup else cue.text
+                    markupText?.let {
+                        TranscriptRegexFilters.extractSpeaker(it.toString(), transcriptFormat)?.let { speaker ->
+                            if (previousSpeaker != speaker) {
+                                append("\n\n$speaker\n\n")
```
Comment (geekygecko) [inline modules/features/player/src/main/java/au/com/shiftyjelly/pocketcasts/player/view/transcripts/TranscriptViewModel.kt:165]:
```suggestion
                                append("\n\n")
                                append(speaker)
                                append("\n\n")
```
Feel free to ignore this but should the appends be separate as the StringBuilder is doing them.

File: modules/features/player/src/main/java/au/com/shiftyjelly/pocketcasts/player/view/transcripts/TranscriptFilter.kt [hunk 3]
```diff
@@ -38,6 +40,30 @@ class TranscriptRegexFilters(private val filters: List<TranscriptFilter>) : Tran
                 RegexFilters.tripleOrMoreEmptyLinesFilter,
             ),
         )
+
+        fun extractSpeaker(cue: String, format: TranscriptFormat?) = when (format) {
+            TranscriptFormat.VTT -> regexMatch(input = cue, pattern = "<v (.+?)>", position = 1)
+            TranscriptFormat.SRT -> regexMatch(input = cue, pattern = "^(.+?):", position = 1)
+            else -> null
+        }
```
Comment (geekygecko) [inline modules/features/player/src/main/java/au/com/shiftyjelly/pocketcasts/player/view/transcripts/TranscriptFilter.kt:48]: I could be overthinking this, but is it bad / slow to be creating the regex pattern every time? How about something like the following:
```kotlin
        private val vttRegex = "<v (.+?)>".toRegex()
        private val srtRegex = "^(.+?):".toRegex()

        fun extractSpeaker(cue: String, format: TranscriptFormat?) = when (format) {
            TranscriptFormat.VTT -> regexMatch(input = cue, regex = vttRegex, position = 1)
            TranscriptFormat.SRT -> regexMatch(input = cue, regex = srtRegex, position = 1)
            else -> null
        }

        private fun regexMatch(input: String, regex: Regex, position: Int = 0): String? {
            return try {
                regex.find(input)?.groupValues?.get(position)
            } catch (e: Exception) {
                null
            }
        }
```

**Review by geekygecko — APPROVED** [review]
This worked well and look amazing.

File: modules/features/player/src/main/java/au/com/shiftyjelly/pocketcasts/player/view/transcripts/TranscriptFilter.kt (hunk 3 above)
Comment (ashiagr) [reply modules/features/player/src/main/java/au/com/shiftyjelly/pocketcasts/player/view/transcripts/TranscriptFilter.kt:63]: I extracted all regex patterns outside in 109b44f and 8a7721c, hopefully it improves performance even more!

File: modules/features/player/src/main/java/au/com/shiftyjelly/pocketcasts/player/view/transcripts/TranscriptViewModel.kt (hunk 2 above)
Comment (ashiagr) [reply modules/features/player/src/main/java/au/com/shiftyjelly/pocketcasts/player/view/transcripts/TranscriptViewModel.kt]: Suggestion applied in 947faac.

**Comment by ashiagr** [comment]
👋 @geekygecko 

I'm reverting support for speaker info in VTT transcripts. I'm having issues including library class duplicates in the release apk in `minifyReleaseWithR8` task.

Also, including library class duplicates may cause issues in code maintenance. I'll try supporting speaker info in VTT transcripts separately.

**Pull request merged**